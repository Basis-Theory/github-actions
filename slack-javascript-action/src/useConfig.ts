import { getInput, exportVariable } from "@actions/core";

export type ConfigType = {
  repository: string;
  version: string;
  release_notes: string;
  author: string;
  type: string;
  action_url: string;
  message_id: string | undefined;
  startedTimestamp: string;
  stoppedTimestamp: string;
  status: string;
  job_status: string | undefined;
  channel: string;
  mention_person: string | undefined;
};

export type GithubContextCommitType = {
  message: string;
};

export type GithubContextType = {
  actor: string;
  sha: string;
  server_url: string;
  repository: string;
  run_id: string;
  event: {
    commits?: GithubContextCommitType[];
    repository: {
      name: string;
    };
    release?: {
      tag_name: string;
      body: string;
      author: {
        login: string;
      };
    };
  };
};

const getVersion = (githubContext: GithubContextType): string =>
  githubContext.event.release
    ? githubContext.event.release.tag_name
    : githubContext.sha.slice(0, 6);

const getReleaseNotes = (githubContext: GithubContextType): string => {
  if (githubContext.event.release) {
    return githubContext.event.release.body;
  }

  if (githubContext.event.commits) {
    return githubContext.event.commits
      .map((commit) => commit.message)
      .join("*");
  }

  return "no release notes";
};

const getAuthor = (githubContext: GithubContextType): string =>
  githubContext.event.release &&
  !githubContext.event.release.author.login.includes("github-actions")
    ? githubContext.event.release.author.login
    : githubContext.actor;

const getDateTime = (): string => {
  return Intl.DateTimeFormat("en", {
    timeZone: "America/Chicago",
    day: "2-digit",
    month: "2-digit",
    year: "numeric",
    hour: "numeric",
    minute: "numeric",
    second: "numeric",
    timeZoneName: "short",
  })
    .format(new Date())
    .replace(/\u202F/g, " ");
};

const getStartedTimestamp = (): string => {
  if (process.env.STARTED_TIMESTAMP) {
    return process.env.STARTED_TIMESTAMP;
  }
  exportVariable("STARTED_TIMESTAMP", getDateTime());

  return getDateTime();
};

const getStoppedTimestamp = (): string => getDateTime();

const useConfig = (): ConfigType => {
  console.log(JSON.stringify(JSON.parse(getInput("github")), null, 2));
  const githubContext: GithubContextType = JSON.parse(getInput("github"));
  const status = getInput("status");
  const type = getInput("type");
  const channel = getInput("channel");
  const mention_person = getInput("mention-person");
  const job_status = process.env.job_status;
  const startedTimestamp = getStartedTimestamp();
  const stoppedTimestamp = getStoppedTimestamp();

  return {
    repository: githubContext.event.repository.name,
    version: getVersion(githubContext),
    release_notes: getReleaseNotes(githubContext),
    author: getAuthor(githubContext),
    action_url: `${githubContext.server_url}/${githubContext.repository}/actions/runs/${githubContext.run_id}`,
    message_id: process.env.SLACK_MESSAGE_ID,
    startedTimestamp,
    stoppedTimestamp,
    status,
    type,
    job_status,
    channel,
    mention_person,
  };
};

export default useConfig;
