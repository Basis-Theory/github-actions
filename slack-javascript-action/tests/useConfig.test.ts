import useConfig, { GithubContextType } from "../src/useConfig";
import { jest } from "@jest/globals";
import * as core from "@actions/core";
jest.mock("@actions/core");
const mockedCore = core as jest.Mocked<typeof core>;

const mockGetInput = (data: any) => {
  mockedCore.getInput.mockClear();
  mockedCore.getInput.mockImplementation(
    (input: string | undefined = ""): any => data[input]
  );
};

let githubData: GithubContextType = {
  repository: "testing",
  run_id: "12",
  server_url: "http://test-repo.com",
  sha: "1234561234567890",
  actor: "drewsue",
  event: {
    commits: [
      {
        message: "test commit 1",
      },
      {
        message: "test commit 2",
      },
    ],
    repository: {
      name: "test-repo",
    },
    release: {
      tag_name: "v42.0.0",
      author: {
        login: "luvi",
      },
      body: "notes!",
    },
  },
};

const coreData: { [key: string]: any } = {
  github: JSON.stringify(githubData),
  status: "done",
  channel: "C1234567890",
  "mention-person": "drewsue",
};

beforeEach(() => {
  process.env.SLACK_MESSAGE_ID = "1234567890";
  process.env.job_status = "success";
  jest.useFakeTimers();
  jest.setSystemTime(new Date(2020, 3, 1));
});

afterEach(() => {
  delete process.env.SLACK_MESSAGE_ID;
  delete process.env.job_status;
  jest.useRealTimers();
});

describe("happy path", () => {
  beforeEach(() => {
    mockGetInput(coreData);
  });

  test("should return the expected config", () => {
    expect(useConfig()).toMatchSnapshot();
  });
});

describe("no release", () => {
  beforeEach(() => {
    const localMockData = { ...githubData };
    delete localMockData.event.release;

    mockGetInput({ ...coreData, github: JSON.stringify(localMockData) });
  });

  test("should return the expected config", () => {
    expect(useConfig()).toMatchSnapshot();
  });
});

describe("no release or commit", () => {
  beforeEach(() => {
    const localMockData = { ...githubData };
    delete localMockData.event.release;
    delete localMockData.event.commits;

    mockGetInput({ ...coreData, github: JSON.stringify(localMockData) });
  });

  test("should return the expected config", () => {
    expect(useConfig()).toMatchSnapshot();
  });
});
