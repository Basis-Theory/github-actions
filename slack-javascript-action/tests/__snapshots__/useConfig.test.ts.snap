// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`happy path should return the expected config 1`] = `
{
  "action_url": "http://test-repo.com/testing/actions/runs/12",
  "author": "luvi",
  "channel": "C1234567890",
  "job_status": "success",
  "mention_person": "drewsue",
  "message_id": "1234567890",
  "release_notes": "notes!",
  "repository": "test-repo",
  "startedTimestamp": "03/31/2020, 7:00:00 PM CDT",
  "status": "done",
  "stoppedTimestamp": "03/31/2020, 7:00:00 PM CDT",
  "type": undefined,
  "version": "v42.0.0",
}
`;

exports[`no release or commit should return the expected config 1`] = `
{
  "action_url": "http://test-repo.com/testing/actions/runs/12",
  "author": "drewsue",
  "channel": "C1234567890",
  "job_status": "success",
  "mention_person": "drewsue",
  "message_id": "1234567890",
  "release_notes": "no release notes",
  "repository": "test-repo",
  "startedTimestamp": "03/31/2020, 7:00:00 PM CDT",
  "status": "done",
  "stoppedTimestamp": "03/31/2020, 7:00:00 PM CDT",
  "type": undefined,
  "version": "123456",
}
`;

exports[`no release should return the expected config 1`] = `
{
  "action_url": "http://test-repo.com/testing/actions/runs/12",
  "author": "drewsue",
  "channel": "C1234567890",
  "job_status": "success",
  "mention_person": "drewsue",
  "message_id": "1234567890",
  "release_notes": "test commit 1*test commit 2",
  "repository": "test-repo",
  "startedTimestamp": "03/31/2020, 7:00:00 PM CDT",
  "status": "done",
  "stoppedTimestamp": "03/31/2020, 7:00:00 PM CDT",
  "type": undefined,
  "version": "123456",
}
`;

exports[`release created by github should return the expected config 1`] = `
{
  "action_url": "http://test-repo.com/testing/actions/runs/12",
  "author": "drewsue",
  "channel": "C1234567890",
  "job_status": "success",
  "mention_person": "drewsue",
  "message_id": "1234567890",
  "release_notes": "test commit 1*test commit 2",
  "repository": "test-repo",
  "startedTimestamp": "03/31/2020, 7:00:00 PM CDT",
  "status": "done",
  "stoppedTimestamp": "03/31/2020, 7:00:00 PM CDT",
  "type": undefined,
  "version": "123456",
}
`;
